# ЕГЭ-2021 по информатике

Вариант № 7408953
Демонстрационная версия ЕГЭ—2021 по информатике.

## Задание 1 № 27398
```
На рисунке схема дорог Н-ского района изображена в виде графа, в таблице содержатся сведения о протяжённости
каждой из этих дорог (в километрах).
П1 П2 П3 П4 П5 П6 П7
П1 9 7
П2 5 11
П3 12
П4 9 5 4 13 15
П5 4 10 8
П6 11 12 13 10
П7 7 15 8
Так как таблицу и схему рисовали независимо друг от друга, то нумерация населённых пунктов в таблице никак не
связана с буквенными обозначениями на графе. Определите, какова протяжённость дороги из пункта Д в пункт Е. В ответе
запишите целое число — так, как оно указано в таблице.  

Решение.
Сопоставим населённые пункты графа и населённые пункты в таблице.
Из Д ведут пять дорог. Только из пункта 4 ведут пять дорог.
Из А ведёт одна дорога. Только из пункта 3 ведёт одна дорога.
Из Б ведёт четыре дороги. Только из пункта 6 ведёт четыре дороги.
Из В ведут две дороги: одна в Д, другая — в Б. Из пункта 2 такжеведут две дороги: одна в Д, другая — в Б,
следовательно, пункт 2 — это пункт В.
Из пункта 1 ведут две дороги — одна в Д, другая в пункт 7, на графе остался только пункт Е, соответствующий
данным условиям. Таким образом, пункт 1 — это пункт Е, а пункт 7 — это пункт К.
Остались только пункты Г и 5, значит, пункт Г и есть пункт 5.
Таким образом, длина дороги из пункта Д в пункт Е равна 9.
```
Ответ: 9.

## Задание 2 № 27399
Логическая функция F задаётся выражением `(x ∨ y) ∧ ¬(y ≡ z) ∧ ¬w`. На рисунке приведён частично заполненный фрагмент таблицы истинности функции F, содержащий неповторяющиеся строки. Определите, какому столбцу таблицы истинности функции F соответствует каждая из переменных x, y, z, w.

| Переменная 1 | Переменная 2 | Переменная 3 | Переменная 4 | Функция |
|:--------------:|:--------------:|:--------------:|:--------------:|:---------:|
| 1 || 1 || 1 |
|0| 1|| 0| 1|
||1| 1| 0| 1|||

В ответе напишите буквы x, y, z, w в том порядке, в котором идут соответствующие им столбцы (сначала — буква, соответствующая первому столбцу; затем — буква, соответствующая второму столбцу, и т. д.). Буквы в ответе пишите подряд, никаких разделителей между буквами ставить не нужно.

**Пример.**  
Пусть задано выражение x → y, зависящее от двух переменных x и y, и фрагмент таблицы истинности:

|Переменная 1| Переменная 2 |Функция|
|:--:|:-:|:-:|
|???| ???| F|
|0| 1| 0|

Тогда первому столбцу соответствует переменная y, а второму столбцу соответствует переменная x. В ответе нужно
написать: yx.  

**Решение.**  
Составим таблицу истинности для выражения (x ∨ y) ∧ ¬(y ≡ z) ∧ ¬w

In [2]:
print('x y z w')
for x in 0, 1:
    for y in 0, 1:
        for z in 0, 1:
            for w in 0, 1:
                if (x or y) and not (y == z) and not w:
                    print(x, y, z, w)

x y z w
0 1 0 0
1 0 1 0
1 1 0 0


Сопоставим эти наборы с приведенным в задании фрагментом таблицы истинности.
Ни в одном из наборов переменная w не принимает единичное значение, следовательно, переменной w соответствует
четвертый столбец таблицы.

| Переменная 1 | Переменная 2 | Переменная 3 | w | Функция |
|:--------------:|:--------------:|:--------------:|:--------------:|:---------:|
|1| |1|0|1|
|0|1| |0|1|
| |1|1|0|1|

Заметим, что в первой и в третьей строках таблицы как минимум две переменные принимают единичные значения,
следовательно, набор (0, 1, 0, 0) может соответствовать только второй строке таблицы, тогда во второй строке в третьем
столбце стоит 0, а второй столбец соответствует переменной у, принимающей в этом наборе единичное значение.

| z | Переменная 2 | Переменная 3 | w | Функция |
|:--------------:|:--------------:|:--------------:|:--------------:|:---------:|
|1| |1|0|1|
|0|1| |0|1|
|0|1|1|0|1|

Заметим, что переменная, стоящая в третьем столбце таблицы, принимает единичное значение дважды, значит, третий
столбец соответствует переменной х.

| z | y | Переменная 3 | w | Функция |
|:--------------:|:--------------:|:--------------:|:--------------:|:---------:|
|1|0|1|0|1|
|0|1| |0|1|
|0|1|1|0|1|

Тогда первый столбец соответствует переменной z.

Ответ: zyxw

## Задание 3 № 27400
Ниже представлены два фрагмента таблиц из базы данных о жителях микрорайона. Каждая строка таблицы 2
содержит информацию о ребёнке и об одном из его родителей. Информация представлена значением поля ID в
соответствующей строке таблицы 1. Определите на основании приведённых данных ID женщины, ставшей матерью в
наиболее молодом возрасте. При вычислении ответа учитывайте только информацию из приведённых фрагментов таблиц.
```
Таблица 1
ID Фамилия_И. О. Пол Год рождения
14 Краснова Н.А. Ж 1937
24 Сканави И.П. М 1943
25 Сканави П.И. М 1974
26 Сканави П.П. М 2001
34 Кущенко А.И. Ж 1964
35 Кущенко В.С. Ж 1990
36 Кущенко С.С. М 1964
44 Лебедь А.С. Ж 1938
45 Лебедь В.А. М 1953
46 Гросс О.С. Ж 1993
47 Гросс П.О. М 2009
54 Клычко А.П. Ж 1995
64 Крот П.А. Ж 1973
... ... ... ...
Таблица 2
ID_Родителя ID_Ребенка
24 25
44 25
25 26
64 26
24 34
44 34
34 35
36 35
14 36
34 46
36 46
25 54
64 54
... ...
```
Решение.
Используя данные таблиц, найдём данные всех матерей и их первых детей (первым идут данные матери, вторым идут
данные первого ребёнка):
1. ID 14, Краснова Н.А., 1937: первый ребёнок — ID 36, Кущенко С.С., 1964, возраст при появлении первого
ребёнка — 27.
2. ID 34, Кущенко А.И., 1964: первый ребёнок — ID 35, Кущенко В.С., 1990, возраст при появлении первого
ребёнка — 26.
3 . ID 44, Краснова Н.А., 1938: первый ребёнок — ID 34, Кущенко А.И., 1964, возраст при появлении первого
ребёнка — 26.
4. ID 64, Крот П.А., 1973: первый ребёнок — ID 54, Клычко А.П., 1995, возраст при появлении первого ребёнка — 22.
Таким образом, у ID 64 в самом молодом возрасте появился первый ребёнок.

Ответ: 64.

## Задание 4 № 27401
Для кодирования некоторой последовательности, состоящей из букв Л, М, Н, П, Р, решили использовать
неравномерный двоичный код, удовлетворяющий условию, что никакое кодовое слово не является началом другого
кодового слова. Это условие обеспечивает возможность однозначной расшифровки закодированных сообщений. Для букв
Л, М, Н использовали соответственно кодовые слова 00, 01, 11. Для двух оставшихся букв — П и Р — кодовые слова
неизвестны.  
Укажите кратчайшее возможное кодовое слово для буквы П, при котором код будет удовлетворять указанному
условию. Если таких кодов несколько, укажите код с наименьшим числовым значением.

Решение.
Для трёх букв кодовые слова уже известны, осталось подобрать для оставшихся двух букв такие кодовые слова,
которые будут являться кратчайшими и удовлетворять условию Фано.
Кодовым словом не могут быть ни 0, ни 1, потому что есть кодовые слова, начинающиеся с 0 и 1. Для буквы П нельзя
использовать кодовое слово 10, поскольку в этом случае нельзя будет закодировать букву Р. Для кодирования букв П и Р
можно использовать кодовые слова 100 и 101. Кратчайшее слово с наименьшим числовым значением — 100.

Ответ: 100.

## Задание 5 № 27402
На вход алгоритма подаётся натуральное число N. Алгоритм строит по нему новое число R следующим образом.
1. Строится двоичная запись числа N.
2. К этой записи дописываются справа ещё два разряда по следующему правилу:  
а) складываются все цифры двоичной записи числа N, и остаток от деления суммы на 2 дописывается в конец числа
(справа). Например, запись 11100 преобразуется в запись 111001;  
б) над этой записью производятся те же действия — справа дописывается остаток от деления суммы её цифр на 2.  

Полученная таким образом запись (в ней на два разряда больше, чем в записи исходного числа N) является двоичной
записью искомого числа R. Укажите такое наименьшее число N, для которого результат работы данного алгоритма больше
числа 77. В ответе это число запишите в десятичной системе счисления.

Решение. Рассмотрим числа, большие 77, и найдем минимальное число, которое является результатом работы алгоритма.
7810 = 100 11102 — является результатом работы алгоритма.
Выполнив обратное преобразование, получим число 100112 = 1910.

Ответ: 19.

In [1]:
N = 0
while 1:
    N += 1
    R = bin(N)[2:]
    R += str(R.count('1') % 2)
    R += str(R.count('1') % 2)
    if int(R, 2) > 77:
        break    
print(f"{N = }")

N = 19


## Задание 6 № 27403
Определите, при каком наибольшем введённом значении переменной s программа выведет число 64. Для Вашего удобства программа представлена на четырёх языках программирования.

    Python

    s = int(input())
    s = s // 10
    n = 1
    while s < 51:
    s = s + 5
    n = n * 2
    print(n)

Ответ: 259.

In [4]:
def print_n(s):
    s = s // 10
    n = 1
    while s < 51:
        s = s + 5
        n = n * 2
    # print(n)
    return n


print(max((s for s in range(1, 1000) if print_n(s) == 64)))

259


## Задание 7 № 27404
Для хранения произвольного растрового изображения размером 128×320 пикселей отведено 20 Кбайт памяти без учёта размера заголовка файла. Для кодирования цвета каждого пикселя используется одинаковое количество бит, коды пикселей записываются в файл один за другим без промежутков. Какое максимальное количество цветов можно использовать в изображении?

Решение.
Объём растрового изображения находится как произведение количества пикселей в изображении на объём памяти x,
необходимый для хранения цвета одного пикселя: $128 · 320 · x < 20 · 2^{13}$ бит, откуда x = 4 бит. Значит, в изображении
можно использовать не более $2^4 = 16$ цветов.

Ответ: 16.

In [5]:
2**((20 * 2**13) / (128 * 320))

16.0

## Задание 8 № 27405
Игорь составляет таблицу кодовых слов для передачи сообщений, каждому сообщению соответствует своё кодовое слово. В качестве кодовых слов Игорь использует трёхбуквенные слова, в которых могут быть только буквы Ш, К, О, Л, А, причём буква К появляется ровно 1 раз. Каждая из других допустимых букв может встречаться в кодовом слове любое количество раз или не встречаться совсем. Сколько различных кодовых слов может использовать Игорь?

Решение.
Пусть К стоит на первом месте. Тогда на остальных двух позициях может стоять любая из четырёх оставшихся букв.
То есть всего `4*4=16` комбинаций.
Если К стоит на втором месте, то также остаётся две позиции, на каждой из которых может находиться любая из
четырёх оставшихся цифр.
Такие же рассуждения, если К стоит на третьем месте.
То есть всего получается `16*3=48` вариантов.

In [6]:
from itertools import product

print(len([word for word in product("ШКОЛА", repeat=3) if word.count("К")==1]))

48


In [26]:
abc = "ШКОЛА"
word_cnt = 0
for a1 in abc: 
    for a2 in abc:
        for a3 in abc:
            word_cnt += (a1+a2+a3).count("К") == 1
print(f"{word_cnt} слов")

48 слов


## Задание 9 № 27406
Откройте файл электронной таблицы, содержащей вещественные числа — результаты ежечасного измерения
температуры воздуха на протяжении трёх месяцев.  
Задание 9  
Найдите разность между максимальным значением температуры и её средним арифметическим значением. В ответе
запишите только целую часть получившегося числа.

Решение.  
Для поиска максимального значения температуры воспользуемся формулой =МАКС(B2:Y92) в ячейке Z2.
Максимальное значения температуры равно 38,0. Теперь в ячейке Z3 с помощью формулы =СРЗНАЧ(B2:Y92) найдём
среднее арифметическое значение всех измерений — 23,7. Теперь найдём разность в ячейке Z4 с помощью формулы =Z2–
Z3: 38,0 − 23,7 = 14,3. Тогда ответ — 14.  
Ответ: 14.

## Задание 10 № 27407
С помощью текстового редактора определите, сколько раз, не считая сносок, встречается слово «долг» или «Долг» в
тексте романа в стихах А. С. Пушкина «Евгений Онегин». Другие формы слова «долг», такие как «долги», «долгами» и
т. д., учитывать не следует. В ответе укажите только число.
Задание 10

Решение.
Воспользуемся поисковыми средствами текстового редактора. В строке поиска последовательно будем вводить
сначала " долг", потом "Долг ". Подсчитав общее количество результатов, получаем ответ — 1.

Ответ: 1.

## Задание 11 № 27408
При регистрации в компьютерной системе каждому пользователю выдаётся пароль, состоящий из 15 символов и
содержащий только символы из 8-символьного набора: А, В, C, D, Е, F, G, H. В базе данных для хранения сведений о
каждом пользователе отведено одинаковое и минимально возможное целое число байт. При этом используют
посимвольное кодирование паролей, все символы кодируют одинаковым и минимально возможным количеством бит.  
Кроме собственно пароля, для каждого пользователя в системе хранятся дополнительные сведения, для чего отведено 24
байт на одного пользователя.  
Определите объём памяти (в байтах), необходимый для хранения сведений о 20 пользователях. В ответе запишите
только целое число — количество байт.

Решение. Согласно условию, в пароле могут быть использованы 8 символов. Известно, что с помощью N бит можно
закодировать $2^N$ различных вариантов. Поскольку $2^3$ = 8, то для записи каждого из 8 символов необходимо 3 бит.
Для хранения всех 15 символов пароля нужно 3 · 15 = 45 бит, а т. к. для записи используется целое число байт, то
берём ближайшее не меньшее значение, кратное восьми, это число 48 = 6 · 8 бит (6 байт).
Для хранения всех сведений об одном пользователе используется 6 + 24 = 30 байт. Таким образом, для хранения
сведений о двадцати пользователях необходимо 30 · 20 = 600 байт.

Ответ: 600.

## Задание 12 № 27409
Исполнитель Редактор получает на вход строку цифр и преобразовывает её. Редактор может выполнять две команды, в
обеих командах v и w обозначают цепочки цифр.

А) заменить (v, w).  
Эта команда заменяет в строке первое слева вхождение цепочки v на цепочку w. Например, выполнение команды заменить (111, 27) преобразует строку 05111150 в строку 0527150. Если в строке нет вхождений цепочки v, то выполнение команды заменить (v, w) не меняет эту строку.  
Б) нашлось (v).  
Эта команда проверяет, встречается ли цепочка v в строке исполнителя Редактор. Если она встречается, то команда возвращает логическое значение «истина», в противном случае возвращает значение «ложь». Строка исполнителя при этом не изменяется.
```
Цикл
ПОКА условие
последовательность команд
КОНЕЦ ПОКА
выполняется, пока условие истинно.
В конструкции
ЕСЛИ условие
ТО команда1
КОНЕЦ ЕСЛИ
выполняется команда1 (если условие истинно).
В конструкции
ЕСЛИ условие
ТО команда1
ИНАЧЕ команда2
КОНЕЦ ЕСЛИ
выполняется команда1 (если условие истинно) или команда2 (если условие ложно).
Какая строка получится в результате применения приведённой ниже программы к строке, состоящей из 70 идущих
подряд цифр 8? В ответе запишите полученную строку.
НАЧАЛО
ПОКА нашлось (2222) ИЛИ нашлось (8888)
ЕСЛИ нашлось (2222)
ТО заменить (2222, 88)
ИНАЧЕ заменить (8888, 22)
КОНЕЦ ЕСЛИ
КОНЕЦ ПОКА
КОНЕЦ
```
Решение.
Данный алгоритм сначала заменит четыре первых восьмёрки на две двойки, на следующем шаге цикла сделает то же
самое, а на третьем шаге цикла заменит четыре получившихся двойки на две восьмёрки. Получаем, что каждые три шага
цикла из последовательности удаляется шесть восьмёрок.Через одиннадцать троек шагов цикла в последовательности
останется четыре восьмёрки. На последнем шаге цикла они будут заменены на две двойки.
Таким образом, получим строку 22.

Ответ: 22.

In [ ]:
s = '8' * 70

while '2222' in s or '8888' in s:
    if '2222' in s:
        s = s.replace('2222', '88', 1)
    else:
        s = s.replace('8888', '22', 1)
        
print(f"{s = }")

s = '22'


## Задание 13 № 27410
На рисунке представлена схема дорог, связывающих города А, Б, В, Г, Д, Е, Ж, З, И, К, Л, М. По каждой дороге можно
двигаться только в одном направлении, указанном стрелкой.
Сколько существует различных путей из города А в город М, проходящих через город В?
```
Решение.
Количество путей до города Х = количество путей добраться в любой из тех городов, из которых есть дорога в Х.
При этом если путь должен не проходить через какой-то город, нужно просто не учитывать этот город при подсчёте
сумм. А если город наоборот обязательно должен лежать на пути, тогда для городов, в которые из нужного города идут
дороги, в суммах нужно брать только этот город.

С помощью этого наблюдения посчитаем последовательно количество путей до каждого из городов:
А = 1
Б = А = 1
Д = А = 1
Г = А + Д = 1 + 1 = 2
В = А + Б + Г = 4
З = В = 4 (Г и Д не учитываем, поскольку путь должен проходить через В)
Ж = В + З = 4 + 4 = 8 (Е не учитываем, поскольку путь должен проходить через В)
И = З + Ж = 4 + 8 = 12
К = Л = И =12
М = К + Л = 24

Примечание. Необходимо найти количество различных путей из города А в город М, проходящих через город В.

Приведем другое решение.
Количество путей из города А в город М, проходящих через город В, равно произведению количества путей из города
А в город В и количества путей из города В в город М.
Найдем количество путей из города А в город В:
А = 1
Б = А = 1
Д = А = 1
Г = А + Д = 1 + 1 = 2
В = А + Б + Г = 4.
Заметим, что из города В в город М можно добраться только через город И. Следовательно, количество путей из
города В в город М равно произведению количества путей из города В в город И и количества путей из города И в город
М.
Найдем количество путей из города В в город И (при этом В — исходный пункт):
В = 1
З = В =1
Ж = В + З = 1 + 1 = 2
И = З + Ж = 1 + 2 = 3.
Из города И в город М есть два пути: И—К—М и И—Л—м.
Таким образом, количество путей из города А в город М, проходящих через город В, равно 4 · 3 · 2 = 24.
```
Ответ: 24.

## Задание 14 № 27411
Значение выражения $49^7 + 7^{21} − 7$ записали в системе счисления с основанием 7.
Сколько цифр «6» содержится в этой записи?

Решение.
Последовательно будем преобразовывать данное выражение:
$49^7 + 7^{21} − 7 = 7^{14} + 7^{21} - 7$  
Это вычитание $7^{14} - 7$ в системе счисления с основанием 7 будет выглядеть как тринадцать шестёрок и один ноль. А
$7^{21}$ как одна единица и 21 ноль.
Таким образом, всего 13 шестёрок.

Ответ: 13.

In [27]:
x = 49**7 + 7**21 - 7

cnt = 0
while x > 0:
    if x % 7 == 6:
        cnt += 1
    x //= 7

print(f"{cnt = }")

cnt = 13


## Задание 15 № 27412

Обозначим через ДЕЛ(n, m) утверждение «натуральное число n делится без остатка на натуральное число m». Для какого наибольшего натурального числа А формула ¬ДЕЛ(x, А) → (ДЕЛ(x, 6) → ¬ДЕЛ(x, 9)) тождественно истинна (то есть принимает значение 1 при любом натуральном значении переменной x)?

Ответ: 18.

In [ ]:
DEL = lambda n, m: n%m == 0

for A in range(1000, 1, -1):
    for x in range(1, 1000):
        formula = (not DEL(x, A)) <= (DEL(x, 6) <= (not DEL(x, 9)))
        if not formula:
            break
    if formula:
        print(A)
        break

18


In [ ]:
DEL = lambda n, m: n%m == 0
formula = lambda x, A: (not DEL(x, A)) <= (DEL(x, 6) <= (not DEL(x, 9)))

print(max((A for A in range(1, 1000) if all((formula(x, A) for x in range(1, 1000))))))

18


## Задание 16 № 27413
Алгоритм вычисления значения функции F(n), где n — натуральное число, задан следующими соотношениями:

    F(n) = 1 при n = 1;
    F(n) = n + F(n − 1), если n — чётно;
    F(n) = 2 × F(n − 2), если n> 1 и при этом n — нечётно.

Чему равно значение функции F(26)?

Ответ: 4122.

In [ ]:
def F(n):
    if n == 1:
        return 1
    if n%2 == 0:
        return n + F(n-1)
    if n>1 and n%2:
        return 2 * F(n-2)

print(F(26))

4122


## Задание 17 № 27414

Рассматривается множество целых чисел, принадлежащих числовому отрезку [1016; 7937], которые делятся на 3 и не
делятся на 7, 17, 19, 27. Найдите количество таких чисел и максимальное из них. В ответе запишите два целых числа без
пробелов и других дополнительных символов: сначала количество, затем максимальное число.
Для выполнения этого задания можно написать программу или воспользоваться редактором электронных таблиц.

Ответ: 15687935.

In [ ]:
num = [i for i in range(1016, 7937+1) if not(i%3) and i%7 and i%17 and i%19 and i%27]

print(len(num), max(num))

1568 7935


## Задание 18 № 27415
Квадрат разлинован на N×N клеток (1 < N < 17). Исполнитель Робот может перемещаться по клеткам, выполняя за
одно перемещение одну из двух команд: вправо или вниз. По команде вправо Робот перемещается в соседнюю правую
клетку, по команде вниз — в соседнюю нижнюю. При попытке выхода за границу квадрата Робот разрушается. Перед
каждым запуском Робота в каждой клетке квадрата лежит монета достоинством от 1 до 100. Посетив клетку, Робот
забирает монету с собой; это также относится к начальной и конечной клетке маршрута Робота.  
Задание 18  
Откройте файл. Определите максимальную и минимальную денежную сумму, которую может собрать Робот, пройдя
из левой верхней клетки в правую нижнюю. В ответ запишите два числа друг за другом без разделительных знаков —
сначала максимальную сумму, затем минимальную.  
Исходные данные представляют собой электронную таблицу размером N×N, каждая ячейка которой соответствует
клетке квадрата.  
Пример входных данных:

|1| 8| 8| 4|
|:-:|:-:|:-:|:-:|
|10| 1| 1| 3|
|1 |3| 12| 2|
|2| 3 |5 |6|

Для указанных входных данных ответом должна быть пара чисел 41 и 22.

Решение. https://youtu.be/ydg4-i3Td-U

## Задание 19 № 27416
Два игрока, Петя и Ваня, играют в следующую игру. Перед игроками лежат две кучи камней. Игроки ходят по
очереди, первый ход делает Петя. За один ход игрок может добавить в одну из куч (по своему выбору) один камень или
увеличить количество камней в куче в два раза . Например, пусть в одной куче 10 камней, а в другой 5 камней; такую
позицию в игре будем обозначать (10, 5). Тогда за один ход можно получить любую из четырёх позиций: (11, 5), (20, 5),
(10, 6), (10, 10). Для того чтобы делать ходы, у каждого игрока есть неограниченное количество камней.  
Игра завершается в тот момент, когда суммарное количество камней в кучах становится не менее 77. Победителем
считается игрок, сделавший последний ход, т. е. первым получивший такую позицию, при которой в кучах будет 77 или
больше камней.  
В начальный момент в первой куче было семь камней, во второй куче — S камней; 1 ≤ S ≤ 69.  
Будем говорить, что игрок имеет выигрышную стратегию, если он может выиграть при любых ходах противника.  
Описать стратегию игрока — значит описать, какой ход он должен сделать в любой ситуации, которая ему может
встретиться при различной игре противника. В описание выигрышной стратегии не следует включать ходы играющего по
этой стратегии игрока, не являющиеся для него безусловно выигрышными, т. е. не являющиеся выигрышными независимо
от игры противника.  
Известно, что Ваня выиграл своим первым ходом после неудачного первого хода Пети. Укажите минимальное
значение S, когда такая ситуация возможна  
Решение.  
Заметим, что игра должна закончиться в 2 хода. Минимальное значение количества камней в обеих кучах, при котором
игра заканчивается — 77. Эта ситуация возможна, например, когда в первой куче 7 камней, а во второй — 70. Значит,
чтобы Ваня мог выиграть своим первым ходом, количество камней во второй куче должно быть ≥35. Поскольку
удваиванием число 35 получить нельзя, после первого хода Пети во второй куче должно получиться 36 камней. Это
возможно при значении S = 18. При таком минимальном значении S Ваня выиграет своим первым ходом после неудачного
хода Пети.  
Ответ:18.

In [ ]:
def move(k, s, p):
    # p- порядковый номер позиции
    if p == 3:
        return k+s >= 77
    return any((move(k+1, s, p+1), move(k, s+1, p+1), move(k*2, s, p+1), move(k, s*2, p+1)))

print(min((s for s in range(1, 69+1) if move(7, s, 1))))

18


## Задание 20 № 27417
Два игрока, Петя и Ваня, играют в следующую игру. Перед игроками лежат две кучи камней. Игроки ходят по
очереди, первый ход делает Петя. За один ход игрок может добавить в одну из куч (по своему выбору) один камень или
увеличить количество камней в куче в два раза . Например, пусть в одной куче 10 камней, а в другой 5 камней; такую
позицию в игре будем обозначать (10, 5). Тогда за один ход можно получить любую из четырёх позиций: (11, 5), (20, 5),
(10, 6), (10, 10). Для того чтобы делать ходы, у каждого игрока есть неограниченное количество камней.
Игра завершается в тот момент, когда суммарное количество камней в кучах становится не менее 77. Победителем
считается игрок, сделавший последний ход, т. е. первым получивший такую позицию, при которой в кучах будет 77 или
больше камней.  
В начальный момент в первой куче было семь камней, во второй куче — S камней; 1 ≤ S ≤ 69.  
Будем говорить, что игрок имеет выигрышную стратегию, если он может выиграть при любых ходах противника.  
Описать стратегию игрока — значит описать, какой ход он должен сделать в любой ситуации, которая ему может
встретиться при различной игре противника. В описание выигрышной стратегии не следует включать ходы играющего по
этой стратегии игрока, не являющиеся для него безусловно выигрышными, т. е. не являющиеся выигрышными независимо
от игры противника.  
Найдите два таких значения S, при которых у Пети есть выигрышная стратегия, причём одновременно выполняются
два условия: — Петя не может выиграть за один ход; — Петя может выиграть своим вторым ходом независимо от того, как будет ходить Ваня.  
Найденные значения запишите в ответе в порядке возрастания без разделительных знаков.

Решение.  
Из хода решения предыдущего задания можно заключить, что значения S 35 и 36 не подходят, поскольку в этом
случае Петя может выиграть своим первым ходом. Рассмотрим значение S = 34. В этом случае Петя своим первым ходом
может добавить в первую кучу один камень и получить кучу (8, 34). После первого хода Вани может возникнуть одна из
четырёх позиций: (9, 34), (8, 35), (16, 34), (8, 68). Во всех случаях Петя удваивает количество камней во второй куче и
выигрывает своим вторым ходом.  
Второе значение S — 31. При S = 31 Петя удваивает количество камней в первой куче и получает позицию (14, 31).
После первого хода Вани может возникнуть одна из четырёх позиций: (15, 31), (14, 32), (28, 31), (14, 62). Во всех случаях
Петя удваивает количество камней во второй куче и выигрывает своим вторым ходом.  
Таким образом, ответ —3134.

Ответ: 3134.

In [ ]:
# кеширование рекурсии
from functools import lru_cache
@lru_cache(None)


def move(k, s, p):
    # p- порядковый номер позиции
    if k+s >= 77 or p>4:
        return p == 4
    if p%2 == 1:
        # Исходная позиция нечетная - ходит Петя
        return any((move(k+1, s, p+1), move(k, s+1, p+1), move(k*2, s, p+1), move(k, s*2, p+1)))
    else:
        # Исходная позиция четная - ходит Ваня
        return all((move(k+1, s, p+1), move(k, s+1, p+1), move(k*2, s, p+1), move(k, s*2, p+1)))

    
print([s for s in range(1, 69+1) if move(7, s, 1)])      

[31, 34]


## Задание 21 № 27418
Два игрока, Петя и Ваня, играют в следующую игру. Перед игроками лежат две кучи камней. Игроки ходят по
очереди, первый ход делает Петя. За один ход игрок может добавить в одну из куч (по своему выбору) один камень или
увеличить количество камней в куче в два раза . Например, пусть в одной куче 10 камней, а в другой 5 камней; такую
позицию в игре будем обозначать (10, 5). Тогда за один ход можно получить любую из четырёх позиций: (11, 5), (20, 5),
(10, 6), (10, 10). Для того чтобы делать ходы, у каждого игрока есть неограниченное количество камней.  
Игра завершается в тот момент, когда суммарное количество камней в кучах становится не менее 77. Победителем
считается игрок, сделавший последний ход, т. е. первым получивший такую позицию, при которой в кучах будет 77 или
больше камней.  
В начальный момент в первой куче было семь камней, во второй куче — S камней; 1 ≤ S ≤ 69.  
Будем говорить, что игрок имеет выигрышную стратегию, если он может выиграть при любых ходах противника.  
Описать стратегию игрока — значит описать, какой ход он должен сделать в любой ситуации, которая ему может
встретиться при различной игре противника. В описание выигрышной стратегии не следует включать ходы играющего по
этой стратегии игрока, не являющиеся для него безусловно выигрышными, т. е. не являющиеся выигрышными независимо
от игры противника.  
Найдите минимальное значение S, при котором одновременно выполняются два условия: — у Вани есть выигрышная стратегия, позволяющая ему выиграть первым или вторым ходом при любой игре Пети; — у Вани нет стратегии, которая позволит ему гарантированно выиграть первым ходом.

Решение.  
Такое значение S — 30. При S = 30 Петя своим первым ходом может получить одну из четырёх позиций: (7, 31), (8, 30),
(14, 30), (7, 60).  
В позиции (7, 60) Ваня удваивает количество камней во второй куче и выигрывает своим первым ходом.  
Из позиций (14, 30) и (7, 31) Ваня может получить позицию (14, 31). В этом случае после второго хода Пети может
возникнуть одна из четырёх позиций: (15, 31), (14, 32), (28, 31), (14, 62). Во всех случаях Ваня удваивает количество
камней во второй куче и выигрывает своим вторым ходом.  
Из позиции (8, 30) Ваня своим первым ходом может получить позицию (16, 30). После второго хода Пети может
возникнуть одна из четырёх позиций: (17, 30), (16, 31), (32, 30), (16, 60). Во всех случаях Ваня удваивает количество
камней во второй куче и выигрывает своим вторым ходом.  
Таким образом, ответ — 30.

Ответ: 30.

Примечание. Докажем, что при S ≤ 29 либо выигрывает Петя своим первым или вторым ходом, либо игра не завершится за
4 хода.  
При S ≤ 7 Петя своим первым ходом может добавить в большую кучу один камень. Тогда, даже если изначально S = 7,
наибольшее количество камней, которое можно получить суммарно в обеих кучах за 4 хода, каждый раз удваивая
большую кучу, равняется 71.  
При 8 ≤ S ≤ 16 Петя может выбрать такую стратегию, которая не позволит победить Ване за один или два хода. Для
этого Петя каждый ход может прибавлять к первой куче один камень. При этом наибольшее суммарное количество камней
в обеих кучах, которое можно получить за 4 хода, равно 9 + 64 = 73.  
При S = 17. Петя первым ходом может получить позицию (7, 18). Из этой позиции Ваня может получит позиции (8, 18),
(7, 19), (14, 18) и (7, 36). В позиции (7, 36) Петя выигрывает своим вторым ходом. В остальных позициях у Пети есть
стратегия, которая позволяет ему получить позиции, из которых Ваня не сможет выиграть своим вторым ходом.  
При 18 ≤ S ≤ 29 Петя может получить позицию (8, S). В этой позиции Петя либо выигрывает своим вторым ходом, либо
у него есть стратегия, которая позволяет ему получить позиции, в которых Ваня не может выиграть своим первым или
вторым ходом.

In [7]:
# кеширование рекурсии
# from functools import lru_cache
# @lru_cache(None)

def move(k, s, p):
    # p- порядковый номер позиции
    if k+s >= 77 or p>5:
        return p==3 or p==5
    if p%2 == 1:
        # Исходная позиция нечетная - ходит Петя
        return all((move(k+1, s, p+1), move(k, s+1, p+1), move(k*2, s, p+1), move(k, s*2, p+1)))
    else:
        # Исходная позиция четная - ходит Ваня
        return any((move(k+1, s, p+1), move(k, s+1, p+1), move(k*2, s, p+1), move(k, s*2, p+1)))


print(min([s for s in range(1, 69+1) if move(7, s, 1)]))

30


In [ ]:
def move(k, s, p):
    # p- порядковый номер позиции
    if k+s >= 77 or p>3:
        return p == 3
    if p%2 == 0:
        # Исходная позиция четная - ходит Ваня
        return any((move(k+1, s, p+1), move(k, s+1, p+1), move(k*2, s, p+1), move(k, s*2, p+1)))
    else:
        # Исходная позиция четная - ходит Ваня
        return all((move(k+1, s, p+1), move(k, s+1, p+1), move(k*2, s, p+1), move(k, s*2, p+1)))

print([s for s in range(1, 69+1) if move(7, s, 1)])
# print('Нет стратегии')

[]


## Задание 22 № 27419
Ниже на пяти языках программирования записан алгоритм. Получив на вход число x, этот алгоритм печатает два
числа: L и M. Укажите наибольшее число x, при вводе которого алгоритм печатает сначала 4, а потом 5.
```
x = int(input())
Q = 9
L = 0
while x >= Q:
L = L + 1
x = x - Q
M = x
if M < L:
M = L
L = x
print(L)
print(M)
```
Решение.  
Данный алгоритм ищет два числа числа: целая часть от деления значения x на 9 и остаток от деления значения x на 9.  
Далее, если остаток от деления числа x на 9 меньше, чем целая часть от этого деления, алгоритм меняет местами значения
L и M.  
Заметим, что в результате работы алгоритма на экран будут выведены два числа: 4 и 5. Значит, чтобы найти
наибольшее число x, число 5 должно являться целой частью от деления x на 9, а число 4 — остатком от деления x на 9.
Таким образом, получаем ответ — x = 9 · 5 + 4 = 49.

Ответ: 49.

In [ ]:
def LM(x):
    
    # x = int(input())
    Q = 9
    L = 0
    while x >= Q:
        L = L + 1
        x = x - Q
    M = x
    if M < L:
        M = L
        L = x
    #print(L)
    #print(M)
    
    return L, M

print(max((x for x in range(1000) if LM(x)[0]==4 and LM(x)[1]==5)))

49


## Задание 23 № 27420
Исполнитель преобразует число на экране.  
У исполнителя есть две команды, которым присвоены номера:

1. Прибавить 1  
2. Умножить на 2

Первая команда увеличивает число на экране на 1, вторая умножает его на 2. Программа для исполнителя — это
последовательность команд.  
Сколько существует программ, для которых при исходном числе 1 результатом является число 20 и при этом
траектория вычислений содержит число 10?  
Траектория вычислений программы — это последовательность результатов выполнения всех команд программы.  
Например, для программы 121 при исходном числе 7 траектория будет состоять из чисел 8, 16, 17.

Решение.  
Искомое количество программ равно произведению количества программ, получающих из числа 1 число 10, на
количество программ, получающих из числа 10 число 20.  
F(n,finish) — количество программ, которые число n преобразуют в число finish.  
Таким образом, количество программ, удовлетворяющих условию задачи равно 14 · 2 = 28.

Ответ: 28.

In [ ]:
def F(n, finish):
    if n == finish:
        return 1
    if n > finish:
        return 0
    return F(n+1, finish) + F(n*2, finish)
    
    
print(F(1, 10) * F(10, 20))

28


## Задание 24 № 27421
Текстовый файл состоит не более чем из $10^6$ символов X, Y и Z. Определите максимальное количество идущих подряд
символов, среди которых каждые два соседних различны.  
Для выполнения этого задания следует написать программу. Ниже приведён файл, который необходимо обработать с
помощью данного алгоритма.  
Задание 24

Решение.  
Для решения данной задачи будем посимвольно считывать текстовый файл. Объявим переменные. Первая нужна для
определения длины каждой последовательности неповторяющихся символов, вторая — для хранения максимальной
длины такой последовательности. Алгоритм будет сравнивать значение текущего символа со значением предыдущего и,
если символы не будут повторяться, увеличивать значения счётчика.

In [25]:
with open('24_demo.txt') as file:
    #line = 'XXZYXXXXXXXYYYYZZZZZYYYYYYXXXXXXXXXXXXYZXZZ'
    line = file.read()

max_cnt = cnt = 1
for i in range(1, len(line)):
    cnt = 1 if line[i] == line[i-1] else cnt+1
    max_cnt = max(max_cnt, cnt)

print(f"{max_cnt = }")

max_cnt = 35


## Задание 25 № 27422
Напишите программу, которая ищет среди целых чисел, принадлежащих числовому
отрезку [174457; 174505], числа, имеющие ровно два различных натуральных делителя, не
считая единицы и самого числа. Для каждого найденного числа запишите эти два делителя
в два соседних столбца на экране с новой строки в порядке возрастания произведения этих
двух делителей. Делители в строке также должны следовать в порядке возрастания.  
Например, в диапазоне [5; 9] ровно два различных натуральных делителя имеют числа 6
и 8, поэтому для этого диапазона вывод на экране должна содержать следующие значения:  
2 3  
2 4  

Ответ:

Решение.  
Решим задачу перебором. Будем проверять количество делителей каждого числа из диапазона, если их количество
равно двум — записываем их в список делителей d.

In [ ]:
for i in range(174457, 174505+1):
    d = [j for j in range(2, i) if i%j == 0]
    if len(d) == 2:
        print(*d)

3 58153
7 24923
59 2957
13 13421
149 1171
5 34897
211 827
2 87251


## Задание 26 № 27423
Системный администратор раз в неделю создаёт архив пользовательских файлов. Однако объём диска, куда он
помещает архив, может быть меньше, чем суммарный объём архивируемых файлов. Известно, какой объём занимает файл
каждого пользователя.  
По заданной информации об объёме файлов пользователей и свободном объёме на архивном диске определите
максимальное число пользователей, чьи файлы можно сохранить в архиве, а также максимальный размер имеющегося
файла, который может быть сохранён в архиве, при условии, что сохранены файлы максимально возможного числа
пользователей.  
Входные данные.  
Задание 26  
В первой строке входного файла находятся два числа: S — размер свободного места на диске (натуральное число, не
превышающее 10 000) и N — количество пользователей (натуральное число, не превышающее 1000). В следующих N
строках находятся значения объёмов файлов каждого пользователя (все числа натуральные, не превышающие 100), каждое в отдельной строке.  
Запишите в ответе два числа: сначала наибольшее число пользователей, чьи файлы могут быть помещены в архив,
затем максимальный размер имеющегося файла, который может быть сохранён в архиве, при условии, что сохранены
файлы максимально возможного числа пользователей.
```
Пример входного файла:
100 4  
80  
30
50
40
```
При таких исходных данных можно сохранить файлы максимум двух пользователей. Возможные объёмы этих двух
файлов 30 и 40, 30 и 50 или 40 и 50. Наибольший объём файла из перечисленных пар — 50, поэтому ответ для
приведённого примера:  
```
2 50
```
Ответ:

> Решение.  
Сначала считаем в массив данные из файла. После этого отсортируем массив в порядке возрастания. Таким образом,
последовательно складывая элементы массива с начала и сравнивая сумму с размером свободного места на диске получим максимальное количество пользователей, чьи файлы могут поместиться на диске. Далее, вычитая из найденной суммы наибольший файл в текущей последовательности, будем пробовать прибавлять файлы с большим весом. Если такой файл будет найден, то заменяем значение наибольшего файла, который возможно поместить на диск.  
В результате работы данного алгоритма при вводе данных из файла в условии получаем ответ — 568 50.

Ответ: 568 50.

*Лучше решать сортировкой в Excel*

In [ ]:
with open('26_demo.txt') as file:
    S, N = map(int, file.readline().split())
    data = sorted([int(file.readline()) for _ in range(N)])

occupied, user = 0, 0
while occupied + data[user] <= S:
    occupied += data[user]
    user += 1

print(user, max((size for size in data if size-data[user-1] <= S-occupied)))

568 50


## Задание 27 № 27424
Имеется набор данных, состоящий из пар положительных целых чисел. Необходимо выбрать из каждой пары ровно
одно число так, чтобы сумма всех выбранных чисел не делилась на 3 и при этом была максимально возможной.
Гарантируется, что искомую сумму получить можно. Программа должна напечатать одно число — максимально
возможную сумму, соответствующую условиям задачи.
```
Входные данные.
Файл A
Файл B
```
Даны два входных файла (файл A и файл B), каждый из которых содержит в первой строке количество пар N
(1 ≤ N ≤ 100000). Каждая из следующих N строк содержит два натуральных числа, не превышающих 10 000.
```
Пример организации исходных данных во входном файле:
6
1 3
5 12
6 9
5 4
3 3
1 1
```
Для указанных входных данных значением искомой суммы должно быть число 32.
В ответе укажите два числа: сначала значение искомой суммы для файла А, затем для файла B.
Предупреждение: для обработки файла B не следует использовать переборный алгоритм, вычисляющий сумму для
всех возможных вариантов, поскольку написанная по такому алгоритму программа будет выполняться слишком долго.

Ответ:

> Решение.  
Последовательно считывая данные из файла, будем прибавлять к сумме максимальное число в паре. Также заметим,
что в случае, если получившееся в результате суммирования максимальных чисел во всех парах число будет кратно трём,
достаточно будет вычесть из этой суммы минимальную разницу между какими-либо двумя числами. Для этого при
считывании пар помимо максимального числа в каждой паре будем искать минимальную разницу среди пар, не кратную
трём.  

In [1]:
sum_max_ab = 0
min_dif = 999_999

with open('27-B_demo.txt') as file:
    for _ in range(int(file.readline())):
        a, b = map(int, file.readline().split())
        sum_max_ab += max(a, b)
        if abs(a-b) % 3:
            min_dif = min(abs(a-b), min_dif)
            
if sum_max_ab % 3 == 0:
    sum_max_ab -= min_dif
print(f"максимальная сумма {sum_max_ab}")

максимальная сумма 399762080


В результате работы данного алгоритма при вводе данных из файла A ответ — 127127, из файла B — 399762080.